chat-android/src/main/java/com/ably/chat/AtomicCoroutineScope.kt

```diff
@@ -0,0 +1,82 @@
+package com.ably.chat
+
+import io.ably.annotation.Experimental
+import java.util.concurrent.PriorityBlockingQueue
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+/**
+ * AtomicCoroutineScope is a thread safe wrapper to run multiple operations mutually exclusive.
+ * All operations are atomic and run with given priority.
+ * Accepts scope as a constructor parameter to run operations under the given scope.
+ * See [Kotlin Dispatchers](https://kt.academy/article/cc-dispatchers) for more information.
+ */
+class AtomicCoroutineScope(private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default)) {
+
+    private val sequentialScope: CoroutineScope = CoroutineScope(Dispatchers.Default.limitedParallelism(1))
+
+    private class Job(
+        private val priority: Int,
+        val coroutineBlock: suspend CoroutineScope.() -> Any,
+        val deferredResult: CompletableDeferred<Any>,
+        val queuedPriority: Int,
+    ) :
+        Comparable<Job> {
+        override fun compareTo(other: Job): Int {
+            if (this.priority == other.priority) {
+                return this.queuedPriority.compareTo(other.queuedPriority)
+            }
+            return this.priority.compareTo(other.priority)
+        }
+    }
+
+    private val jobs: PriorityBlockingQueue<Job> = PriorityBlockingQueue() // Accessed from both sequentialScope and async method
+    private var isRunning = false // Only accessed from sequentialScope
+    private var queueCounter = 0 // Only accessed from synchronized async method
+
+    /**
+     * @param priority Defines priority for the operation execution.
+     * @param coroutineBlock Suspended function that needs to be executed mutually exclusive under given scope.
+     */
+    @Synchronized
```

Comment:
Why we use synchronized here?

Reply:
Sure!
Actually, I wanted to explore how the async method behaves when called from multiple concurrent coroutines. To do this, I added a [test case](https://github.com/ably-labs/ably-chat-kotlin/blob/95a6a2c6584965930b6da851deffa65d888c9df0/chat-android/src/test/java/com/ably/chat/AtomicCoroutineScopeTest.kt#L94) to cover this scenario. This is a practical scenario because `AtomicCoroutineScope.async` can be called from coroutines running on different threads. My goal was to ensure that a thread should add a job at a time, while incrementing the `queueCounter`.

If we made `queueCounter` atomic, we'd face concurrency issues with how jobs are added to the `jobs: PriorityBlockingQueue` and how the `queueCounter: AtomicInteger` is updated. This could lead to different outputs depending on how locks are acquired on these two separate data structures. To address this, I marked the method as synchronized, which ensures that the lock is obtained on both data structures simultaneously. This also eliminated the need to declare `queueCounter` as an `AtomicInteger`. I have added code comments for the same.

I want to emphasize that the implementation was driven entirely by the requirements of the test case. I’ve made sure to cover all relevant scenarios in the tests.

Reply:
Actually, PriorityBlockingQueue is accessed from both the sequentialScope thread and the async method thread. I have already added a comment regarding this [here](https://github.com/ably-labs/ably-chat-kotlin/blob/7405d9c01f0b50b75ce1e91776a0aaccf9845778/chat-android/src/main/java/com/ably/chat/AtomicCoroutineScope.kt#L36).

Our current tests using a normal `PriorityQueue` result in flaky behavior due to race conditions caused by read/write operations from two threads.

Reply:
Also, `finishedProcessing` method checks for empty `PriorityBlockingQueue`. So, this is also a read operation from a different thread.